### DISCUSSION TITLE
Help with IndexProvider

### DISCUSSION THREAD

**Discussion by vitalybrandes** [OrchardCMS/OrchardCore#7762] [Q&A] [answered]
I need to create index for ContentItem once its saved from the admin UI.
I create 

```C#
 public class UserByPrinterIndexProvider : IndexProvider<ContentItem>
    {
        public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<UserByPrinterIndex>()
             .Map(contentItem =>
             {
                 var printer = contentItem.As<Printer>();

                 if (printer == null)
                 {
                     return null;
                 }

                 var results = new List<UserByPrinterIndex>();

                 //for each selected printer create index
                 foreach (string user in printer.Content.Printer.Users.UserNames)
                 {
                     results.Add(new UserByPrinterIndex
                     {
                         PrinterId = printer.PrinterId.Text,
                         PrinterContentItemId = contentItem.ContentItemId,
                         UserName = user,
                     });
                 }
                 return results;
             });
        }
    }
```

How cat i run it once content item created, updated, deleted?

**Comment by deanmarcussen** [comment]
Did you register it with dependency injection?

`services.AddSingleton<IIndexProvider, UserByPrinterIndexProvider >();`

That should be all you need to do.

**Reply by vitalybrandes** [reply]
> Did you register it with dependency injection?
> 
> `services.AddSingleton<IIndexProvider, UserByPrinterIndexProvider >();`
> 
> That should be all you need to do.

Yes, I did that 😩.
I feel that I miss something...
Do I need to create driver or something?
How indexprovider executed if I save the content item (Printer)  via Admin Publish Content item btn?

**Comment by jtkech** [comment]
Did you create your index table through a `DataMigration`, as done e.g. by the `OC.Autoroute` module in its `Migrations.cs` file?

**Reply by vitalybrandes** [reply]
Yes I did.

```C#
  public int Create()
        {
            SchemaBuilder.CreateMapIndexTable<UserByPrinterIndex>(table => table
                 .Column<string>(nameof(UserByPrinterIndex.PrinterId))
                 .Column<string>(nameof(UserByPrinterIndex.PrinterContentItemId))
                 .Column<string>(nameof(UserByPrinterIndex.UserName))
             );
            return 1;
        }
```

**Comment by jtkech** [comment]
Did you register it ;)

            services.AddScoped<IDataMigration, Migrations>();

**Reply by vitalybrandes** [reply]
I did and now i even see that that table was created but not populated.

**Comment by jtkech** [comment, marked as answer]
No luck ;)

Need to leave soon, but can you share your full code so that someone will review it

**Reply by vitalybrandes** [reply]
@jtkech 

Its problem to share the code :-(
but ill try to explain here what i did:

At one of my modules I create migration file : 
```C#
 public class PrinterByCompaniesMigration : DataMigration
    {
        public int Create()
        {
            SchemaBuilder.CreateMapIndexTable<PrinterByCompaniesIndex>(table => table
                 .Column<string>(nameof(PrinterByCompaniesIndex.Company))
                 .Column<string>(nameof(PrinterByCompaniesIndex.CompanyContentItemId))
                 .Column<string>(nameof(PrinterByCompaniesIndex.PrinterId))
                 .Column<string>(nameof(PrinterByCompaniesIndex.PrinterContentItemId))
             );
            return 1;
        }
}
```
The table create in db successfully!
as well i created PrinterByCompaniesIndexProvider

```C#
   public class PrinterByCompaniesIndex : MapIndex
    {
        public string Company { get; set; }
        public string CompanyContentItemId { get; set; }
        public string PrinterId { get; set; }
        public string PrinterContentItemId { get; set; }
    }
```

```C#
    public class PrinterByCompaniesIndexProvider : IndexProvider<ContentItem>
    {
        public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<PrinterByCompaniesIndex>()
             .Map(contentItem =>
             {
                 var company = contentItem.As<Company>();

                 if (company == null)
                 {
                     return null;
                 }
                 var results = new List<PrinterByCompaniesIndex>();
                 foreach (var printer in company.Content.Printer)
                 {
                     results.Add(new PrinterByCompaniesIndex
                     {
                         Company = company.CompanyName.Text,
                         CompanyContentItemId = company.ContentItem.ContentItemId,
                         PrinterId = printer.ContentItem.DisplayText,
                         PrinterContentItemId = printer.ContentItem.ContentItemId,
                     });
                 }
                 return results;
             });
        }
    }
```
Both registered in starup.cs

```C#
            services.AddSingleton<IIndexProvider, PrinterByCompaniesIndexProvider>();
            services.AddScoped<IDataMigration, PrinterByCompaniesMigration>();
```

What i expect is, while i am creating, updating or deleting item from **admin** : Content -> ContentItems page
that this index will be updated.
This not happen :-(

**Reply by jtkech** [reply]
First, can you put a breakpoint to see if it is executed

Then if `contentItem.As<Company>()` is not null

So i assume that `Company` is a content part

Oops just saw your question

> Do I need to create driver or something?

Yes or another advanced way to weld / apply your part data to a content item and save it, but the regular way is to implement and register a contentpart specifying the driver that needs to be used, and let the content management coordinate all the things, about parts / drivers / handlers there are many examples in the source code, how there are registered and used.

Currently how are you saving some `Company` data?

Finally if `company.Content.Printer` is not an empty collection

**Reply by vitalybrandes** [reply]
I found the problem, than you!
